source/opt/fix_storage_class.cpp

```diff
@@ -0,0 +1,133 @@
+// Copyright (c) 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "fix_storage_class.h"
+
+#include "source/opt/instruction.h"
+#include "source/opt/ir_context.h"
+
+namespace spvtools {
+namespace opt {
+
+Pass::Status FixStorageClass::Process() {
+  bool modified = false;
+
+  get_module()->ForEachInst([this, &modified](Instruction* inst) {
+    if (inst->opcode() == SpvOpVariable) {
+      std::vector<Instruction*> uses;
+      get_def_use_mgr()->ForEachUser(
+          inst, [&uses](Instruction* use) { uses.push_back(use); });
+      for (Instruction* use : uses) {
+        modified |= PropagateStorageClass(
+            use, static_cast<SpvStorageClass>(inst->GetSingleWordInOperand(0)));
+      }
+    }
+  });
+  return modified ? Status::SuccessWithChange : Status::SuccessWithoutChange;
```

Comment:
Should there be a warning/error/failure if a storage class mismatch was detected but could not be fixed?  Or does this pass not detect this case?

Reply:
This pass does not check if it fails to fix something.  It would require more code to check if it fails.  Like comparing function parameters for type mismatches or looking at every operand of an OpPhi to make sure the types all match.  This is code that is not needed for the pass, and already exists in the validator.

Running the validator will identify anything that has not been fixed.  I did not feel the need to make it part of this pass.